src/sqlfluff/rules/L052.py

```diff
@@ -193,21 +193,38 @@ def _eval(self, context: RuleContext) -> Optional[LintResult]:
                     anchor_segment = self._handle_trailing_inline_comments(
                         context, anchor_segment
                     )
-
-                    fixes = [
-                        LintFix.replace(
-                            anchor_segment,
-                            [
+                    fixes = []
+                    if anchor_segment == context.segment:
+                        fixes.append(
+                            LintFix.replace(
                                 anchor_segment,
-                                NewlineSegment(),
-                                SymbolSegment(raw=";", type="symbol", name="semicolon"),
-                            ],
-                        ),
-                        LintFix.delete(
-                            context.segment,
-                        ),
-                    ]
```

Comment:
Usually anchor_segment != context.segment (since there was always a meta segment between the semicolon and the end of a select statment). Therefore these 2 fixes were anchored to different segments.

In the case highlighted in the linked issue, there were no meta statements between the semicolon and the closing bracket so anchor_segment = context.segment. This makes sense in the context of other cases in the rule and the gives the correct anchoring behaviour for the lint. But in this case we are trying to apply to LintFixes to the same anchor, which due to [this](https://github.com/sqlfluff/sqlfluff/blob/741e814a0bd6c10592d4683af0337f012dea3b1f/src/sqlfluff/core/parser/segments/base.py#L1014-L1016) line of code gets consumed on the replace and therefore the original semicolon never gets deleted, leading to the observed multiple semicolon behaviour.

In a way, the original logic isn't really wrong. The delete fix is just getting thrown away due to that linked TODO.

Reply:
In the event that `anchor_segment == context.segment` we can apply the fix in one go so added a clause to handle that and avoid the overlapping fix anchors mentioned above.